core/src/transport/global_only.rs

```diff
@@ -0,0 +1,355 @@
+        (a.segments()[0] & 0xfe00) == 0xfc00
+    }
+
+    /// Returns [`true`] if this address is `fe80::/10`.
+    pub const fn is_unicast_link_local(a: Ipv6Addr) -> bool {
+        (a.segments()[0] & 0xffc0) == 0xfe80
+    }
+
+    pub fn is_global(a: Ipv6Addr) -> bool {
+        !(a.is_loopback()
+            || a.is_unspecified()
+            || Self::is_ipv4_mapped_address(a)
+            || Self::is_ipv4_ipv6_translat(a)
+            || Self::is_discard_only_address_block(a)
+            || Self::is_ietf_protocol_assignments(a)
+            || Self::is_teredo(a)
+            || Self::is_port_control_protocol_anycast(a)
+            || Self::is_traversal_relays_nat_anycast(a)
+            || Self::is_benchmarking(a)
+            || Self::is_amt(a)
+            || Self::is_as112_v6(a)
+            || Self::is_deprecated_previously_orchid(a)
+            || Self::is_orchid_v2(a)
+            || Self::is_drone_remote(a)
+            || Self::is_documentation(a)
+            || Self::is_6to4(a)
+            || Self::is_direct_delegation_as112_service(a)
+            || Self::is_unique_local(a)
+            || Self::is_unicast_link_local(a))
+    }
+}
+
+impl<T> GlobalIpOnly<T> {
+    pub fn new(transport: T) -> Self {
+        GlobalIpOnly { inner: transport }
+    }
+}
+
+impl<T: Transport + Unpin> Transport for GlobalIpOnly<T> {
+    type Output = <T as Transport>::Output;
+    type Error = <T as Transport>::Error;
+    type ListenerUpgrade = <T as Transport>::ListenerUpgrade;
+    type Dial = <T as Transport>::Dial;
+
+    fn listen_on(&mut self, addr: Multiaddr) -> Result<ListenerId, TransportError<Self::Error>> {
+        self.inner.listen_on(addr)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        self.inner.remove_listener(id)
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        match addr.iter().next() {
+            Some(Protocol::Ip4(a)) => {
+                if Ipv4Global::is_global(a) {
+                    self.inner.dial(addr)
+                } else {
+                    debug!("Not dialing non global IP address {:?}.", a);
+                    Err(TransportError::MultiaddrNotSupported(addr))
+                }
+            }
+            Some(Protocol::Ip6(a)) => {
+                if Ipv6Global::is_global(a) {
+                    self.inner.dial(addr)
+                } else {
+                    debug!("Not dialing non global IP address {:?}.", a);
+                    Err(TransportError::MultiaddrNotSupported(addr))
+                }
+            }
+            _ => {
+                debug!("Not dialing unsupported Multiaddress {:?}.", addr);
+                Err(TransportError::MultiaddrNotSupported(addr))
+            }
+        }
+    }
+
+    fn dial_as_listener(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<Self::Dial, TransportError<Self::Error>> {
+        self.inner.dial_as_listener(addr)
```

Comment:
@mxinden Shouldn't we check the same conditions here?

Reply:
Yes please. This was an oversight in my kademlia-exporter implementation.

Reply:
@thomaseizinger @mxinden Done